src/Fields/reduced_field.jl

```diff
@@ -137,4 +137,4 @@ ReducedField(loc::Tuple, args...; kwargs...) = ReducedField(loc..., args...; kwa
 reduced_location(loc; dims) = Tuple(i ∈ dims ? Nothing : loc[i] for i in 1:3)
 
 Adapt.adapt_structure(to, reduced_field::ReducedField{X, Y, Z}) where {X, Y, Z} =
-    ReducedField{X, Y, Z}(adapt(to, reduced_field.data), nothing, reduced_field.dims, nothing)
+    ReducedField{X, Y, Z}(adapt(to, reduced_field.data), adapt(to, reduced_field.grid), reduced_field.dims, nothing)
```

Comment:
Whats the error?

Reply:
Ah I see, it was just `nothing` before 👍 

Right now our `grid` are all `isbits` so `adapt` isn't needed; however we will need this in the future when the `grid` has arrays.

Reply:
As a side note --- we often include the `grid` as an explicit argument in kernels, which is why it's often not needed when `adapt`ing. Are we changing this pattern in this PR?

Reply:
I did it in response to this error: https://buildkite.com/clima/oceananigans/builds/1533#bf41fda1-28f0-43f6-870a-13ed10e59f97/39-326

Reply:
But upon closer inspection maybe the issue is that 

```julia
ReducedField{X, Y, Z}(adapt(to, reduced_field.data), nothing, reduced_field.dims, nothing)
```

is calling the "friendly constructor" that validates input when we want to call the low-level constructor via

```julia
ReducedField{X, Y, Z, typeof(adapt(to, reduced_field.data)), Nothing, typeof(reduced_field.dims), Nothing}(adapt(to, reduced_field.data), nothing, reduced_field.dims, nothing)
```
?

Reply:
Yes! I think you want to define a new inner constructor for `ReducedField` that takes in just `data` and `dims` and sets `grid` and `boundary_conditions` to `nothing`:

```julia
    function ReducedField{X, Y, Z}(data::A, dims) where {X, Y, Z, A}
        N = length(dims)
        return new{X, Y, Z, A, Nothing, N, Nothing}(data, nothing, dims, nothing)
    end
```

Inside `struct ReducedField`...

Reply:
@glwagner and @ali-ramadhan - hmm I added the constructor, but it looks like it is still erroring at ``cosd`` for Δxᶜᶠᵃ

```

[1] overdub at /data5/glwagner/.julia-1550/packages/Cassette/Wjztv/src/overdub.jl:595
--
  | [2] cosd at special/trig.jl:1050
  | [3] Δxᶜᶠᵃ at /net/ocean/home/data44/data5/glwagner/.buildkite-agent/builds/sverdrup-11/clima/oceananigans/src/Operators/areas_and_volumes.jl:112
  | [4] macro expansion at /data5/glwagner/.julia-1550/packages/KernelAbstractions/mKsXc/src/extras/loopinfo.jl:26
  |  

```

on GPU.  Maybe I should try just using a regular field with ``Nz=1``? 

Error log is here ( https://buildkite.com/clima/oceananigans/builds/1550#53750651-86fc-424a-a920-7fbaa386f286/37-321 ), could it be something else the an Adapt thing? Something with ``cosd`` maybe?

Reply:
Here is the function signature that seems to annoy things

```
  InvalidIRError: compiling kernel gpu__compute_vertically_integrated_lateral_face_areas!(Cassette.Context{nametype(CUDACtx),KernelAbstractions.CompilerMetadata{KernelAbstractions.NDIteration.StaticSize{(360, 360)},KernelAbstractions.NDIteration.DynamicCheck,Nothing,Nothing,KernelAbstractions.NDIteration.NDRange{2,KernelAbstractions.NDIteration.StaticSize{(23, 23)},KernelAbstractions.NDIteration.StaticSize{(16, 16)},Nothing,Nothing}},Nothing,KernelAbstractions.var"##PassType#253",Nothing,Cassette.DisableHooks}, typeof(Oceananigans.Models.HydrostaticFreeSurfaceModels.gpu__compute_vertically_integrated_lateral_face_areas!), RegularLatitudeLongitudeGrid{Float64,Bounded,Bounded,Bounded,OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}, NamedTuple{(:Ax, :Ay),Tuple{ReducedField{Face,Center,Nothing,OffsetArray{Float64,3,CuDeviceArray{Float64,3,1}},RegularLatitudeLongitudeGrid{Float64,Bounded,Bounded,Bounded,OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},1,Nothing},ReducedField{Center,Face,Nothing,OffsetArray{Float64,3,CuDeviceArray{Float64,3,1}},RegularLatitudeLongitudeGrid{Float64,Bounded,Bounded,Bounded,OffsetArray{Float64,1,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},1,Nothing}}}) resulted in invalid LLVM IR
```

Maybe something is still weird it seems to have ```KernelAbstractions.NDIteration.StaticSize{(16, 16)}``` - but not sure it should if this is a reduced field. Entity on RHS is not a ReduucedField - maybe that is causing some issue?